### PR TITLE
Features/#414 unused log file in use section

### PR DESCRIPTION
**Pull request by dakochik**
Issue #414 is resolved. Now `SmkUnusedLogFileInspection` is checks `use rule` too.

### CODE REVIEW

File: src/main/kotlin/com/jetbrains/snakecharm/inspections/SmkUnusedLogFileInspection.kt [hunk 1]
```diff
@@ -28,21 +27,22 @@ class SmkUnusedLogFileInspection : SnakemakeInspection() {
...
                         else -> {
                             return@forEach
                         }
                     }
                 }
                 if (resolveResult != null) {
-                    resolveResult as SmkRule
+                    resolveResult as SmkRuleOrCheckpoint
                     val ruleLogSection = resolveResult.getSectionByName(SnakemakeNames.SECTION_LOG)
                     val message =
```
Comment (iromeo) [inline src/main/kotlin/com/jetbrains/snakecharm/inspections/SmkUnusedLogFileInspection.kt:47]: Local returns like `return@forEach` are hard to track in complex code. Please extract foreach method lambda into a separate method

File: src/main/kotlin/com/jetbrains/snakecharm/inspections/SmkUnusedLogFileInspection.kt [hunk 2]
```diff
@@ -28,21 +27,22 @@ class SmkUnusedLogFileInspection : SnakemakeInspection() {
 
         override fun visitSmkUse(use: SmkUse) {
             val logSection = use.getSectionByName(SnakemakeNames.SECTION_LOG) ?: return
-            use.getOverriddenRuleReferences()?.forEach {
+            use.getImportedRuleNames()?.children?.forEach {
```
Comment (iromeo) [inline src/main/kotlin/com/jetbrains/snakecharm/inspections/SmkUnusedLogFileInspection.kt]: Change log not updated

**Review by iromeo — CHANGES_REQUESTED** [review]
1) Additionally quick fix on :
image

results in an exception:
```
java.util.NoSuchElementException: List is empty.
	at kotlin.collections.CollectionsKt___CollectionsKt.last(_Collections.kt:416)
	at com.jetbrains.snakecharm.inspections.SmkUnusedLogFileInspection$CreateShellSectionWithLogReference.invoke(SmkUnusedLogFileInspection.kt:170)
	at com.intellij.codeInspection.LocalQuickFixOnPsiElement.applyFix(LocalQuickFixOnPsiElement.java:100)
	at com.intellij.codeInspection.LocalQuickFixOnPsiElement.applyFix(LocalQuickFixOnPsiElement.java:90)
	at com.intellij.codeInspection.LocalQuickFixOnPsiElement.applyFix(LocalQuickFixOnPsiElement.java:22)
	at com.intellij.codeInspection.ex.QuickFixWrapper.invoke(QuickFixWrapper.java:78)
	at com.intellij.codeInsight.intention.impl.IntentionActionWithTextCaching$MyIntentionAction.lambda$invoke$0(IntentionActionWithTextCaching.java:209)
	at com.intellij.util.SlowOperations.lambda$allowSlowOperations$0(SlowOperations.java:77)
	at com.intellij.util.SlowOperations.allowSlowOperations(SlowOperations.java:68)
	at com.intellij.util.SlowOperations.allowSlowOperations(SlowOperations.java:76)
	at com.intellij.codeInsight.intention.impl.IntentionActionWithTextCaching$MyIntentionAction.invoke(IntentionActionWithTextCaching.java:209)
	at com.intellij.codeInsight.intention.impl.ShowIntentionActionsHandler.lambda$invokeIntention$4(ShowIntentionActionsHandler.java:251)
	at com.intellij.openapi.application.WriteAction.lambda$run$1(WriteAction.java:105)
	at com.intellij.openapi.application.impl.ApplicationImpl.runWriteActionWithClass(ApplicationImpl.java:947)
	at com.intellij.openapi.application.impl.ApplicationImpl.runWriteAction(ApplicationImpl.java:973)
	at com.intellij.openapi.application.WriteAction.run(WriteAction.java:104)
	at com.intellij.codeInsight.intention.impl.ShowIntentionActionsHandler.invokeIntention(ShowIntentionActionsHandler.java:251)
	at com.intellij.codeInsight.intention.impl.ShowIntentionActionsHandler.lambda$chooseActionAndInvoke$3(ShowIntentionActionsHandler.java:227)
	at com.intellij.openapi.command.impl.CoreCommandProcessor.executeCommand(CoreCommandProcessor.java:216)
	at com.intellij.openapi.command.impl.CoreCommandProcessor.executeCommand(CoreCommandProcessor.java:172)
	at com.intellij.openapi.command.impl.CoreCommandProcessor.executeCommand(CoreCommandProcessor.java:162)
	at com.intellij.openapi.command.impl.CoreCommandProcessor.executeCommand(CoreCommandProcessor.java:148)
	at com.intellij.codeInsight.intention.impl.ShowIntentionActionsHandler.chooseActionAndInvoke(ShowIntentionActionsHandler.java:226)
	at com.intellij.codeInsight.intention.impl.IntentionListStep.lambda$applyAction$3(IntentionListStep.java:135)
	at com.intellij.ide.IdeEventQueue.onFocusEvent(IdeEventQueue.java:569)
	at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$9(IdeEventQueue.java:486)
	at com.intellij.openapi.application.impl.ApplicationImpl.runIntendedWriteActionOnCurrentThread(ApplicationImpl.java:781)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:502)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)
```
Maybe it is due to wrong parsing of rule w/o body (see syntax error)

2) if module_414.smk is
```
rule rule_414_foo:
    input: ""
   # shell: "echo {log}"  # <-  FP here

rule rule_414_foo2:
    input: ""
   # shell: "echo {log}"  # <-  FP here
 ```

quick fix fixes only first rule which is not consistent:
```
rule rule_414_foo:
    input: ""
    shell: "echo TODO >{log} 2>&1"
   # shell: "echo {log}"  # <-  FP here

rule rule_414_foo2:
    input: ""
   # shell: "echo {log}"  # <-  FP here

```
here better to clarify rule name in `Add 'shell' section...` and show 2 quick fixes, or maybe do not show it at all.

**Comment by iromeo** [comment]
After merge with master, quick fix still fails for rule w/o sections:
```
java.util.NoSuchElementException: List is empty.
	at kotlin.collections.CollectionsKt___CollectionsKt.last(_Collections.kt:416)
	at com.jetbrains.snakecharm.inspections.SmkUnusedLogFileInspection$CreateShellSectionWithLogReference.invoke(SmkUnusedLogFileInspection.kt:170)
	at com.intellij.codeInspection.LocalQuickFixOnPsiElement.applyFix(LocalQuickFixOnPsiElement.java:100)
	at com.intellij.codeInspection.LocalQuickFixOnPsiElement.applyFix(LocalQuickFixOnPsiElement.java:90)
	at com.intellij.codeInspection.LocalQuickFixOnPsiElement.applyFix(LocalQuickFixOnPsiElement.java:22)
	at com.intellij.codeInspection.ex.QuickFixWrapper.invoke(QuickFixWrapper.java:78)
	at com.intellij.codeInsight.intention.impl.IntentionActionWithTextCaching$MyIntentionAction.lambda$invoke$0(IntentionActionWithTextCaching.java:209)
	at com.intellij.util.SlowOperations.lambda$allowSlowOperations$0(SlowOperations.java:77)
	at com.intellij.util.SlowOperations.allowSlowOperations(SlowOperations.java:68)
	at com.intellij.util.SlowOperations.allowSlowOperations(SlowOperations.java:76)
	at com.intellij.codeInsight.intention.impl.IntentionActionWithTextCaching$MyIntentionAction.invoke(IntentionActionWithTextCaching.java:209)
	at com.intellij.codeInsight.intention.impl.ShowIntentionActionsHandler.lambda$invokeIntention$4(ShowIntentionActionsHandler.java:251)
	at com.intellij.openapi.application.WriteAction.lambda$run$1(WriteAction.java:105)
	at com.intellij.openapi.application.impl.ApplicationImpl.runWriteActionWithClass(ApplicationImpl.java:947)
	at com.intellij.openapi.application.impl.ApplicationImpl.runWriteAction(ApplicationImpl.java:973)
	at com.intellij.openapi.application.WriteAction.run(WriteAction.java:104)
	at com.intellij.codeInsight.intention.impl.ShowIntentionActionsHandler.invokeIntention(ShowIntentionActionsHandler.java:251)
	at com.intellij.codeInsight.intention.impl.ShowIntentionActionsHandler.lambda$chooseActionAndInvoke$3(ShowIntentionActionsHandler.java:227)
	at com.intellij.openapi.command.impl.CoreCommandProcessor.executeCommand(CoreCommandProcessor.java:216)
	at com.intellij.openapi.command.impl.CoreCommandProcessor.executeCommand(CoreCommandProcessor.java:172)
	at com.intellij.openapi.command.impl.CoreCommandProcessor.executeCommand(CoreCommandProcessor.java:162)
	at com.intellij.openapi.command.impl.CoreCommandProcessor.executeCommand(CoreCommandProcessor.java:148)
	at com.intellij.codeInsight.intention.impl.ShowIntentionActionsHandler.chooseActionAndInvoke(ShowIntentionActionsHandler.java:226)
	at com.intellij.codeInsight.intention.impl.IntentionListStep.lambda$applyAction$3(IntentionListStep.java:135)
	at com.intellij.ide.IdeEventQueue.onFocusEvent(IdeEventQueue.java:569)
	at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$9(IdeEventQueue.java:486)
	at com.intellij.openapi.application.impl.ApplicationImpl.runIntendedWriteActionOnCurrentThread(ApplicationImpl.java:781)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:502)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)
	```

File: src/main/kotlin/com/jetbrains/snakecharm/inspections/SmkUnusedLogFileInspection.kt (hunk 1 above)
Comment (dakochik) [reply src/main/kotlin/com/jetbrains/snakecharm/inspections/SmkUnusedLogFileInspection.kt]: Additional method `handleUseReference` is created

File: src/main/kotlin/com/jetbrains/snakecharm/inspections/SmkUnusedLogFileInspection.kt [hunk 3]
```diff
@@ -28,21 +27,22 @@ class SmkUnusedLogFileInspection : SnakemakeInspection() {
...
+                        is SmkRuleOrCheckpoint -> break
                         is SmkReferenceExpression -> reference = resolveResult
                         else -> {
                             return@forEach
                         }
                     }
                 }
                 if (resolveResult != null) {
-                    resolveResult as SmkRule
+                    resolveResult as SmkRuleOrCheckpoint
```
[the remark this answers is not in the document]
Comment (dakochik) [reply src/main/kotlin/com/jetbrains/snakecharm/inspections/SmkUnusedLogFileInspection.kt]: Changed to `resolveResult is SmkRuleOrCheckpoint`

**Comment by dakochik** [comment]
Now `SmkUnusedLogFileInspection` doesn't throw an exception. Quick fix is shown for each rule, which doesn't contain a log reference in execution section.

**Comment by iromeo** [comment]
Code generation isn't nice for the case:

Before
```
rule rule_414_foo:
    # input: ""
   # shell: "echo {log}"  # <-  FP here

rule rule_414_foo2:
    # input: ""
   # shell: "echo {log}"  # <-  FP here

```

After:
```
rule rule_414_foo: shell: "echo TODO >{log} 2>&1"
    # input: ""
   # shell: "echo {log}"  # <-  FP here

rule rule_414_foo2:
    # input: ""
   # shell: "echo {log}"  # <-  FP here
```
In the previous impl it was ok:
```
rule rule_414_foo:
    shell: "echo TODO >{log} 2>&1"
    # input: ""
   # shell: "echo {log}"  # <-  FP here

rule rule_414_foo2:
    # input: ""
   # shell: "echo {log}"  # <-  FP here
```

**Comment by dakochik** [comment]
If target rule for quick fix has an empty body, the `log` section indent will be used as indent for inserted `shell` section.
Benefits: `log` indent always exists if quick fix is offered, in contrast to comments indentation which may not exist or may have wrong indentation.
Drawbacks: the `use rule` section with `log` subsection may has indentation which is different to indentation of target rule. But maybe it's a rare use case. Example:
```
rule a:

if (True):
    use rule a as b with:
        log: "log.log"
```

**Pull request merged**